File: scripts/adjust-classifications.js
#!/usr/bin/env node

const fs = require('fs').promises;
const path = require('path');

// 需要从工具类移除的内容标题列表
const titlesToRemove = [
  '调qing一点通 – 晚熟青年',
  '法律相关电子书',
  '全国独家研学手册',
  '冷眼观爱 七天找到女朋友',
  '生财宝典（100位生财高手的判断路径和经验）',
  '怎样打开自学之门',
  '完全图解恋爱心理学',
  '如何让你爱的人爱上你',
  '欧阳春晓：6周维密薄肌线条弹力带普拉提',
  '如何利用下班时间写作',
  '步某非烟第1-4季 合集',
  '大学同学提升幸福感实用教程'
];

async function backupFile(filePath) {
  const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
  const backupPath = `${filePath}.backup-${timestamp}`;
  const content = await fs.readFile(filePath, 'utf8');
  await fs.writeFile(backupPath, content);
  console.log(`✅ 备份创建成功: ${backupPath}`);
  return backupPath;
}

async function extractToolsToRemove(toolsFilePath) {
  const content = await fs.readFile(toolsFilePath, 'utf8');
  const removedTools = [];
  const lines = content.split('\n');

  let currentTool = [];
  let isInTool = false;
  let braceCount = 0;
  let shouldRemove = false;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    // 检测工具对象的开始
    if (line.includes('{') && !isInTool) {
      isInTool = true;
      currentTool = [line];
      braceCount = (line.match(/{/g) || []).length - (line.match(/}/g) || []).length;

      // 检查是否包含要移除的标题
      if (line.includes('title:')) {
        for (const title of titlesToRemove) {
          if (line.includes(title)) {
            shouldRemove = true;
            break;
          }
        }
      }
    } else if (isInTool) {
      currentTool.push(line);

      // 检查标题
      if (line.includes('title:')) {
        for (const title of titlesToRemove) {
          if (line.includes(title)) {
            shouldRemove = true;
            break;
          }
        }
      }

      // 更新括号计数
      braceCount += (line.match(/{/g) || []).length - (line.match(/}/g) || []).length;

      // 工具对象结束
      if (braceCount === 0) {
        if (shouldRemove) {
          removedTools.push(currentTool.join('\n'));
          console.log(`📋 提取工具: ${currentTool.find(l => l.includes('title:'))?.trim()}`);
        }

        currentTool = [];
        isInTool = false;
        shouldRemove = false;
      }
    }
  }

  return removedTools;
}

async function removeToolsFromFile(toolsFilePath) {
  const content = await fs.readFile(toolsFilePath, 'utf8');
  const lines = content.split('\n');
  const newLines = [];

  let isInTool = false;
  let braceCount = 0;
  let shouldRemove = false;
  let skipTool = false;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    if (line.includes('{') && !isInTool) {
      isInTool = true;
      braceCount = (line.match(/{/g) || []).length - (line.match(/}/g) || []).length;

      // 向前查看是否包含要移除的标题
      for (const title of titlesToRemove) {
        if (content.slice(i, i + 1000).includes(title)) {
          shouldRemove = true;
          skipTool = true;
          break;
        }
      }

      if (!skipTool) {
        newLines.push(line);
      }
    } else if (isInTool) {
      if (!skipTool) {
        newLines.push(line);
      }

      braceCount += (line.match(/{/g) || []).length - (line.match(/}/g) || []).length;

      if (braceCount === 0) {
        isInTool = false;
        shouldRemove = false;
        skipTool = false;

        // 如果刚移除了一个工具，检查下一行是否是逗号
        if (skipTool && i + 1 < lines.length && lines[i + 1].trim() === ',') {
          i++; // 跳过逗号
        }
      }
    } else {
      newLines.push(line);
    }
  }

  // 清理多余的逗号
  let cleanedContent = newLines.join('\n');
  cleanedContent = cleanedContent.replace(/,(\s*),/g, ','); // 移除连续逗号
  cleanedContent = cleanedContent.replace(/,(\s*)\]/g, '\n]'); // 移除数组末尾的逗号

  await fs.writeFile(toolsFilePath, cleanedContent);
  console.log(`✅ 已从 tools.ts 中移除 ${titlesToRemove.length} 个内容`);
}

async function saveRemovedTools(removedTools) {
  const outputPath = path.join(__dirname, '..', 'removed-tools.json');
  await fs.writeFile(outputPath, JSON.stringify(removedTools, null, 2));
  console.log(`💾 移除的工具已保存到: ${outputPath}`);
  console.log('   可以手动将这些内容转换为教程格式后添加到 tutorials.ts');
}

async function main() {
  try {
    console.log('🚀 开始批量调整内容分类...\n');

    const toolsFilePath = path.join(__dirname, '..', 'src', 'data', 'tools.ts');

    // 1. 备份原文件
    console.log('📦 步骤 1: 备份原文件');
    await backupFile(toolsFilePath);

    // 2. 提取要移除的工具
    console.log('\n📋 步骤 2: 提取要移除的内容');
    const removedTools = await extractToolsToRemove(toolsFilePath);

    // 3. 保存移除的内容
    console.log('\n💾 步骤 3: 保存移除的内容');
    await saveRemovedTools(removedTools);

    // 4. 从 tools.ts 中移除内容
    console.log('\n✂️ 步骤 4: 从 tools.ts 中移除内容');
    await removeToolsFromFile(toolsFilePath);

    console.log('\n✨ 分类调整完成！');
    console.log('\n📝 下一步建议：');
    console.log('1. 检查 removed-tools.json 文件');
    console.log('2. 将内容转换为教程格式');
    console.log('3. 添加到 tutorials.ts 文件');
    console.log('4. 运行 npm run dev 测试变更');

  } catch (error) {
    console.error('❌ 错误:', error.message);
    process.exit(1);
  }
}

main();

File: scripts/adjust_classifications.py
#!/usr/bin/env python3

import re
import json
import shutil
from datetime import datetime
from pathlib import Path

# 需要从工具类移除的内容标题列表
TITLES_TO_REMOVE = [
    '调qing一点通 – 晚熟青年',
    '法律相关电子书',
    '全国独家研学手册',
    '冷眼观爱 七天找到女朋友',
    '生财宝典（100位生财高手的判断路径和经验）',
    '怎样打开自学之门',
    '完全图解恋爱心理学',
    '如何让你爱的人爱上你',
    '欧阳春晓：6周维密薄肌线条弹力带普拉提',
    '如何利用下班时间写作',
    '步某非烟第1-4季 合集',
    '大学同学提升幸福感实用教程'
]

def backup_file(file_path):
    """备份文件"""
    timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
    backup_path = f"{file_path}.backup_{timestamp}"
    shutil.copy2(file_path, backup_path)
    print(f"✅ 备份创建成功: {backup_path}")
    return backup_path

def extract_tool_objects(content):
    """提取所有工具对象"""
    tools = []

    # 找到 tools 数组的开始和结束
    match = re.search(r'export const tools:\s*Tool\[\]\s*=\s*\[(.*?)\];', content, re.DOTALL)
    if not match:
        raise ValueError("无法找到 tools 数组")

    tools_content = match.group(1)

    # 使用更智能的方式解析对象
    current_obj = []
    brace_count = 0
    in_string = False
    string_char = None
    in_template = False

    for i, char in enumerate(tools_content):
        # 处理字符串
        if not in_string and not in_template and char in ["'", '"']:
            in_string = True
            string_char = char
        elif in_string and char == string_char and tools_content[i-1] != '\\':
            in_string = False
            string_char = None

        # 处理模板字符串
        if not in_string and char == '`':
            in_template = not in_template

        # 只在非字符串中计算括号
        if not in_string and not in_template:
            if char == '{':
                if brace_count == 0:
                    current_obj = []
                brace_count += 1
            elif char == '}':
                brace_count -= 1

        current_obj.append(char)

        # 完成一个对象
        if brace_count == 0 and current_obj and current_obj[0] == '{':
            obj_str = ''.join(current_obj)
            # 提取标题
            title_match = re.search(r'title:\s*[\'"`]([^\'"`]+)[\'"`]', obj_str)
            if title_match:
                tools.append({
                    'title': title_match.group(1),
                    'content': obj_str.strip()
                })
            current_obj = []

    return tools

def filter_tools(tools, titles_to_remove):
    """分离要保留和要移除的工具"""
    tools_to_keep = []
    tools_to_remove = []

    for tool in tools:
        if tool['title'] in titles_to_remove:
            tools_to_remove.append(tool)
            print(f"  ✂️ 移除: {tool['title']}")
        else:
            tools_to_keep.append(tool)

    return tools_to_keep, tools_to_remove

def rebuild_tools_file(original_content, tools_to_keep):
    """重建 tools.ts 文件"""
    # 重建工具数组内容
    tools_content = ',\n  '.join([tool['content'] for tool in tools_to_keep])

    # 替换原有的 tools 数组
    new_content = re.sub(
        r'export const tools:\s*Tool\[\]\s*=\s*\[(.*?)\];',
        f'export const tools: Tool[] = [\n  {tools_content}\n];',
        original_content,
        flags=re.DOTALL
    )

    return new_content

def save_removed_tools(tools_to_remove, output_path):
    """保存移除的工具"""
    # 转换为更易读的格式
    removed_data = []
    for tool in tools_to_remove:
        # 尝试提取更多信息
        description_match = re.search(r'description:\s*[\'"`]([^\'"`]+)[\'"`]', tool['content'])
        category_match = re.search(r'category:\s*[\'"`]([^\'"`]+)[\'"`]', tool['content'])

        removed_data.append({
            'title': tool['title'],
            'description': description_match.group(1) if description_match else '',
            'category': category_match.group(1) if category_match else '',
            'original_content': tool['content']
        })

    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(removed_data, f, ensure_ascii=False, indent=2)

    print(f"✅ 已保存 {len(removed_data)} 个移除的工具到: {output_path}")

def main():
    """主函数"""
    print("🚀 开始批量调整内容分类...\n")

    # 设置路径
    script_dir = Path(__file__).parent
    project_root = script_dir.parent
    tools_file = project_root / 'src' / 'data' / 'tools.ts'

    if not tools_file.exists():
        print(f"❌ 错误: 找不到文件 {tools_file}")
        return

    try:
        # 1. 备份原文件
        print("📦 步骤 1: 备份原文件")
        backup_file(tools_file)

        # 2. 读取文件内容
        print("\n📖 步骤 2: 读取并解析文件")
        with open(tools_file, 'r', encoding='utf-8') as f:
            original_content = f.read()

        # 3. 提取所有工具对象
        print("\n🔍 步骤 3: 提取工具对象")
        all_tools = extract_tool_objects(original_content)
        print(f"  找到 {len(all_tools)} 个工具")

        # 4. 分离要保留和移除的工具
        print("\n✂️ 步骤 4: 识别要移除的内容")
        tools_to_keep, tools_to_remove = filter_tools(all_tools, TITLES_TO_REMOVE)

        # 5. 重建文件内容
        print("\n🔨 步骤 5: 重建文件内容")
        new_content = rebuild_tools_file(original_content, tools_to_keep)

        # 6. 保存修改后的文件
        print("\n💾 步骤 6: 保存修改")
        with open(tools_file, 'w', encoding='utf-8') as f:
            f.write(new_content)

        # 7. 保存移除的内容
        if tools_to_remove:
            removed_file = project_root / 'removed-tools-detailed.json'
            save_removed_tools(tools_to_remove, removed_file)

        # 统计信息
        print("\n✨ 分类调整完成！")
        print(f"📊 统计:")
        print(f"  - 原有工具: {len(all_tools)} 个")
        print(f"  - 移除工具: {len(tools_to_remove)} 个")
        print(f"  - 保留工具: {len(tools_to_keep)} 个")

        if len(tools_to_remove) < len(TITLES_TO_REMOVE):
            not_found = set(TITLES_TO_REMOVE) - set([t['title'] for t in tools_to_remove])
            print(f"\n⚠️  未找到以下标题:")
            for title in not_found:
                print(f"    - {title}")

        print("\n📝 下一步建议：")
        print("1. 检查 removed-tools-detailed.json 文件")
        print("2. 将内容转换为教程格式")
        print("3. 添加到 tutorials.ts 文件")
        print("4. 运行 npm run dev 测试变更")

    except Exception as e:
        print(f"❌ 错误: {e}")
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    main()